<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>R crash course: A quick introduction to R</title>
    <meta charset="utf-8" />
    <meta name="author" content="Alex Sanchez-Pla" />
    <script src="libs/header-attrs-2.11/header-attrs.js"></script>
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/metropolis-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="my.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# R crash course: A quick introduction to R
### Alex Sanchez-Pla
### Departamento de Genetica, Microbiologa y Estadistica (UB) <br> Unidad de Estadistica y Bioinformatica (VHIR)
### Version 2021-11-17

---


class: inverse, middle, center

name: Introduction

# Introduction
---

# A Crash Course in R. Outline

- Why R
    + R basics
    + How does one work with R and Rstudio
- Getting Started
    + A primer of data import
    + Variables and data types
    + Functions, Packages and more stuff
- Working with data
    + _Selecting, Filtering and ordering datasets_
    + _A primer of statistics and plots_
    + R Notebooks and RMarkdown

---

# Motivation

- We (you) all work with data, most of the time and often we need to do "things" with those data.

  - I have three lists of genes and I would like to see which genes they have in common (or which  ones appear only in one list).
  - We have received the data from that lab but I only want to work with a subset of the samples.
  - Is it possible to repeat that plot changing the line colors, the font size etc?
  - I have some scripts tu re-run an analysis but I don't know how to start

- These, and many other things can be done with a basic knowledge of R.

---

# [What is R?](https://www.r-project.org/about.html)

- R is a _language and environment_ for statistical computing and graphics.

- R provides a wide variety of statistical and graphical techniques, and is highly extensible.

- It can be used fro simple tasks to highly complex reproducible projects.

- It compiles and runs on a wide variety of UNIX platforms and similar systems Windows and MacOS.

---

# R PRO's (why you are here!)

- The system is
    + free (as in _free beer_)
    + It's platform independent
    + It is constantly improving (2 new versions/year)

- It is a statistical tool
    + Implements almost every statistical method that exists
    + Great graphics (Examples)
    + Simple reporting tools
    + Also state-of-the-art in Bioinformatics through the [Bioconductor Project](http://bioconductor.org).

- Programming language
    + Easy to automate repetitive tasks (Example_1.1)
    + Possibility to create user friendly web interfaces  with a moderate effort. (Examples)

---

# R CON's

- R is mainly used issuing commands from a console
    + less user friendly than almost any other statistical tool you may know.

- Constantly having new versions may affect our projects

- Not necessarily the best language nor suitable for every existing task

---

# How is R used

- Different ways to use R, but the best trade-off simplicity-efficiency is provided by Rstudio

&lt;img src="images/RStudio.png" width="2560" /&gt;

&lt;!-- \begin{figure} --&gt;
&lt;!-- \includegraphics[width=0.85\linewidth]{images/RStudio.png} --&gt;
&lt;!-- \end{figure} --&gt;

---

# Exercise

- Get to know R. Visit the R-project page and see what can be found there.

- If you haven't done it before, download and install R and Rstudio in your computer

- Open R studio. Look at the panels and figgure out what can we do at each window.

---

class: inverse, middle, center

name: UsingR

# Using R.

---

# Commands, Objects and Functions

- Shortly, using R consists of

    + Working with *objects* using *commands* and *functions*


&lt;img src="images/basicConcepts.png" width="1139" /&gt;
&lt;!-- \begin{figure} --&gt;
&lt;!-- \includegraphics[width=0.85\linewidth]{images/basicConcepts.png} --&gt;
&lt;!-- \end{figure} --&gt;

---

# Variables and data types

- Data managed in R ...
    + is stored as *variables*

.pull-left[
- Variables can be of distinct _types_
    + Numerical
        - numeric (13.7)
        - int (3)
    + Character
        - "R is cute"
    + Factors
        - A,B,C,D
        - WT, Mut
]
.pull-right[
- Variables can be contained in distinct _structures_
    + vectors
    + matrices
    + data.frames
    + lists
    + tibble
    + or specific classes that combine multiple structures such as
      - Bioconductor's expressionSet or summarizedExperiment
]
---

# R packages

- R can be used for many different types of data processing and analysis from distinct fields, besides statistics such as Ecology, Omics Sciences, Psychology etc.

- All these capabilities are not present from the begining because most of them will never be used by most users.

- Instead, thay can be added when needed by

  - _installing_ and
  - _loading_ the appropriate packages.

---

# Installing and loading packages

- Imagine we want to analyze some data using cox proportional hazards model.
- A colleague has provided us with some code:


```r
res.cox &lt;- coxph(Surv(time, status) ~ sex, data = lung)
```
`Error in coxph(Surv(time, status) ~ sex, data = lung) : could not find function "coxph"`

- It has not worked because the needed functions were not available
- We need to install and load the package before we can use it.


```r
install.packages("survival")
library(survival)
res.cox &lt;- coxph(Surv(time, status) ~ sex, data = lung)
```

---

# Bioconductor

.pull-left[

- Packages analyse all kinds of Genomic data (&gt;800)
- Compulsory documentation (*vignettes*) for each package
- 6-month release cycle
- [Course Materials](http://bioconductor.org/help/course-materials/)
- [Example data](http://bioconductor.org/packages/release/BiocViews.html#___ExperimentData) and [workflows](http://bioconductor.org/help/workflows/)
- Common, re-usable framework and functionality
- [Available Support](https://support.bioconductor.org/)
    + Often you will be able to interact with the package maintainers / developers and other power-users of the project software
]
.pull-right[

&lt;img src="images/Bioconductor.png" width="1329" /&gt;
]
---

# The `tidyverse`

- The tidyverse is an opinionated collection of R packages designed for data science.
- All packages share an underlying design philosophy, grammar, and data structures.
- The complete tidyverse collection can be installed with:

```r
install.packages("tidyverse")
```
- [https://www.tidyverse.org/](https://www.tidyverse.org/)

---

class: inverse, middle, center

name: EnteringData

# Getting data into R

---

# Importing data with Rstudio

- The easiest way to get data into R is to click on the `ÃŒmport Datasets` button.
- Alternatively R code can be written using functions from `Base R` or the `tidyverse`
    + `Base R` functions start with `read.`: `read.table`, `read.csv`
    + `tidyverse` functions start with `read_`: `read_delim`, `read_csv` or `read_excel`

.center[
&lt;img align = "center" src = "images/RImportData.png" width = "80%"&gt;
]

---

# Working with projects

- Files can be read from any location, let it be a physical support or a web site.
- The simplest and best way to control file location and modularity of your analyses is to create an Rstudio project for each new analysis.
  - Easy way to keep together your data, code and results.
  - Increases portability (avoids forgetting a file in an external folder).
  - I opens the door to infinite possibilities when you learn to _clone_ github projects.

.center[
&lt;img align = "center" src = "images/RStudioProjects.png" width = "60%"&gt;
]

---

# Reading Excel or csv files

- R allows importing any type of dataset, _either with "base" packages or using additional ones.

- An easy way to learn how to import a datset is to do this using the `import`menu and then to check the resulting R code.
after importing file `TIO2+PTYR-human-MSS+MSIvsPD.XLSX` the following code has been created (it appears in the console)


```r
library(readxl)
TIO2_PTYR_human_MSS_MSIvsPD &lt;- read_excel("datasets/TIO2+PTYR-human-MSS+MSIvsPD.XLSX")
View(TIO2_PTYR_human_MSS_MSIvsPD)
```

---

# Exercise

- Repeat the import process using the differnt files contained in the `datasets.zip` file

- Can you tell the differences between the files you have imported?

- What is the type of file "Data2HM"

---

class: inverse, middle, center

name: ManagingData

# Managing and exploring data

---

# Data Exploration with R

- Once a dataset is available it is easy to "have a look at it"


```r
phosphoProtData&lt;- read_excel("Session1.2-IntroRandPracticum/datasets/TIO2+PTYR-human-MSS+MSIvsPD.XLSX")
head(phosphoProtData)
str(phosphoProtData)
summary (phosphoProtData)
```

- Do it by yourselves and notice that categorical variables have been read as characters.

---

class: inverse, middle, center

name: RMarkdown

# Dynamic output with Rmarkdown

---

# Reproducible research with R notebooks

- R and Rstudio are strongly involved in promoting [reproducibility](https://en.wikipedia.org/wiki/Reproducibility) and [reproducible research](https://en.wikipedia.org/wiki/Reproducibility#Reproducible_research).
- This is implemented in **R notebooks**
- A notebook combines
    + Natural language text, e.g. describing what we are doing in our own words.
    + R code with the instructions needed to do the data management or the analysis.
    + The output of the analysis

---

# Creating Notebooks

- A notebook can be created in Rstudio with
    + `File --&gt; New File --&gt; R Notebook`
- The notebook contains example text and code so it is straightforwoard to adapt it to your analysis.
- To produce an html file with text, code and output:
    - Press the button "Preview"
    - Or Select "Knitr to Html"

---

class: inverse, middle, center

name: Resources

# Resources and exercises

---

# Introductory materials

The web is full of all types of materials about R

Below there are a couple of brief introductions:

- [A short introduction to R]( https://stats.idre.ucla.edu/r/seminars/short-introduction-to-r/)

- [Getting started with R](https://github.com/saghirb/Getting-Started-in-R)

---

# Exercise

- Select a dataset with which you wish to work along the course.
- Read it into R
    + How many variables are there in it
    + What are their types
- Try to summarize it briefly
- Create an R notebook to encapsulate all your steps and share it with somebody.
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
